### DISCUSSION TITLE
Analog clock

### DISCUSSION THREAD

**Discussion by xiasi0** [HASwitchPlate/openHASP#679] [Show and tell]
![1710688793815](https://github.com/HASwitchPlate/openHASP/assets/129178072/40c80c1e-2bb6-49af-9665-df3025827b06)

```json
{"page":1,"comment":" ---- 第4页时针 ----"}
{"page":4,"id":1,"obj":"gauge","x":90,"y":90,"w":300,"h":300,"min":0,"max":720,"val":0,"critical_value":400,"label_count":0,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":15,"line_rounded10":1,"pad_inner":-30,"text_font":30,"angle":360,"rotation":180,"hidden":0,"line_width60":0,"line_color10":"#00FF00","bg_opa10":0}
{"page":1,"comment":" ---- 第4页分针 ----"}
{"page":4,"id":2,"obj":"gauge","x":40,"y":40,"w":400,"h":400,"min":0,"max":59,"val":30,"critical_value":80,"label_count":0,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":10,"line_rounded10":1,"pad_inner":-10,"text_font":30,"angle":360,"rotation":177,"hidden":0,"line_width60":0,"line_color10":"#FF0000","bg_opa10":0}

{"page":1,"comment":" ---- 第4页小时刻度 ----"}
{"page":4,"id":3,"obj":"linemeter","x":15,"y":15,"w":450,"h":450,"min":1,"max":12,"val":0,"line_count":13,"angle":360,"bg_opa":0,"hidden":0,"scale_width":20,"hidden":0,"border_width":0,"scale_end_color60":"#FFFFFF"}
{"page":1,"comment":" ---- 第4页分钟刻度 ----"}
{"page":4,"id":4,"obj":"linemeter","x":15,"y":15,"w":450,"h":450,"min":1,"max":36,"val":0,"line_count":49,"angle":360,"bg_opa":0,"rotation":5,"scale_width":8,"hidden":0,"border_width":0,"scale_end_color60":"#FFFFFF"}

{"page":1,"comment":" ---- 第4页小时数字 ----"}
{"page":4,"id":5,"obj":"gauge","x":40,"y":40,"w":400,"h":400,"min":1,"max":12,"val":1,"critical_value":30,"label_count":12,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":0,"line_rounded10":1,"pad_inner":0,"text_font":30,"angle":330,"rotation":195,"hidden":0,"line_width60":0,"bg_opa10":255}
```

**Comment by flexiti** [comment]
Thank for this @xiasi0  👍 
I made a small change for myself and added custom code to make it working

```
{"page":9,"comment":" ---- Hours ----"}
{"page":9,"id":1,"obj":"gauge","x":90,"y":90,"w":250,"h":250,"min":0,"max":720,"val":0,"critical_value":400,"label_count":0,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":15,"line_rounded10":1,"pad_inner":-30,"text_font":30,"angle":360,"rotation":180,"hidden":0,"line_width60":0,"line_color10":"#000000","bg_opa10":0}

{"page":9,"comment":" ---- Minutes ----"}
{"page":9,"id":2,"obj":"gauge","x":40,"y":40,"w":350,"h":350,"min":0,"max":59,"val":10,"critical_value":80,"label_count":0,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":10,"line_rounded10":1,"pad_inner":-10,"text_font":30,"angle":360,"rotation":177,"hidden":0,"line_width60":0,"line_color10":"#000000","bg_opa10":0}

{"page":9,"comment":" ---- Seconds ----"}
{"page":9,"id":3,"obj":"gauge","x":40,"y":40,"w":350,"h":350,"min":0,"max":59,"val":10,"critical_value":80,"label_count":0,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":5,"line_rounded10":1,"pad_inner":-10,"text_font":30,"angle":360,"rotation":177,"hidden":0,"line_width60":0,"line_color10":"#FF0000","bg_opa10":0}


{"page":9,"comment":" ---- hours scale ----"}
{"page":9,"id":4,"obj":"linemeter","x":15,"y":15,"w":400,"h":400,"min":1,"max":12,"val":0,"line_count":13,"angle":360,"bg_opa":0,"hidden":0,"scale_width":20,"hidden":0,"border_width":0,"scale_end_color60":"#FFFFFF"}

{"page":9,"comment":" ---- minutes scale ----"}
{"page":9,"id":5,"obj":"linemeter","x":15,"y":15,"w":400,"h":400,"min":1,"max":36,"val":0,"line_count":49,"angle":360,"bg_opa":0,"rotation":5,"scale_width":8,"hidden":0,"border_width":0,"scale_end_color60":"#FFFFFF"}

{"page":9,"comment":" ---- Numbers ----"}
{"page":9,"id":6,"obj":"gauge","x":40,"y":40,"w":350,"h":350,"min":1,"max":12,"val":1,"critical_value":30,"label_count":12,"line_count":0,"border_width":0,"bg_opa":0,"line_width10":0,"line_rounded10":1,"pad_inner":0,"text_font":20,"angle":330,"rotation":195,"hidden":0,"line_width60":0,"bg_opa10":255}
```
 MyClockS, MyClockM, MyClockH - actual time
in custom_every_second() 
```
{
lv_obj_t* MyObjS= hasp_find_obj_from_page_id(9, 3);  //second hand
            if(MyObjS) 
                  {
                    lv_gauge_set_value(MyObjS, 0, MyClockS); 
                    lv_obj_t* MyObjM= hasp_find_obj_from_page_id(9, 2);  //minute hand
                       if(MyObjM) 
                        {
                         lv_gauge_set_value(MyObjM, 0, MyClockM);  
                         lv_obj_t* MyObjH= hasp_find_obj_from_page_id(9, 1);  //hours hand
                           if(MyObjH) 
                             {
                              int tmpClock =(MyClockH*60) + MyClockM;
                              if (MyClockH >=12)                              //only for 24 hours clock
                                tmpClock = ((MyClockH-12)*60)+ MyClockM;      //only for 24 hours clock
                             lv_gauge_set_value(MyObjH, 0, tmpClock);  
                             }
                        }  
                  }
}
```
@fvanroie 
 Is it possible to create variables (like  "template":"%s" for the label), so that the clock can run locally without custom code??

**Reply by fvanroie** [reply]
There is no arithmetic or scripting language in openHASP that can facilitate coding.
You'll need to use custom code, or develop an [analog clock](https://forum.lvgl.io/t/how-to-create-a-clock/2015) widget that can be parameterized by openHASP like any other LVGL widget.

**Comment by vmfs1968** [comment]
{
lv_obj_t* MyObjS= hasp_find_obj_from_page_id(9, 3);  //second hand
            if(MyObjS) 
                  {
                    lv_gauge_set_value(MyObjS, 0, MyClockS); 
                    lv_obj_t* MyObjM= hasp_find_obj_from_page_id(9, 2);  //minute hand
                       if(MyObjM) 
                        {
                         lv_gauge_set_value(MyObjM, 0, MyClockM);  
                         lv_obj_t* MyObjH= hasp_find_obj_from_page_id(9, 1);  //hours hand
                           if(MyObjH) 
                             {
                              int tmpClock =(MyClockH*60) + MyClockM;
                              if (MyClockH >=12)                              //only for 24 hours clock
                                tmpClock = ((MyClockH-12)*60)+ MyClockM;      //only for 24 hours clock
                             lv_gauge_set_value(MyObjH, 0, tmpClock);  
                             }
                        }  
                  }
}
Can anyone help where can I put this code?

**Reply by flexiti** [reply]
![clk](https://github.com/HASwitchPlate/openHASP/assets/11538048/327f31ee-3111-4bac-a0dc-d07501c9c28a)
source: eg. custom_code.cpp  and compile all

**Reply by vmfs1968** [reply]
thanks
Is it in here?
I tried to compile and got this error

**Reply by mvturnho** [reply]
Would you be willing to share the code for the clock?

**Reply by cpertel** [reply]
Hi
Is there a "public available" solution?
If so, please be so kind and share it.
Thanks a lot.

Regards 
CP

**Comment by vmfs1968** [comment]
![image](https://github.com/HASwitchPlate/openHASP/assets/41055800/a9de4e3a-46fe-467f-9692-2439aa121cdb)

thanks
Is it in here?
I tried to compile and got this error

Converting .pio/build/guition-jc4827w543c_4MB/en.json.gz.txt.o
Converting .pio/build/guition-jc4827w543c_4MB/pages.jsonl.txt.o
Compiling .pio/build/guition-jc4827w543c_4MB/src/custom/clock_code.cpp.o
src/custom/clock_code.cpp:1:1: error: expected unqualified-id before '{' token
 {
 ^
Compiling .pio/build/guition-jc4827w543c_4MB/src/custom/my_custom_fan_template.cpp.o
*** [.pio/build/guition-jc4827w543c_4MB/src/custom/clock_code.cpp.o] Error 1
